files/xml.py

```diff
@@ -0,0 +1,470 @@
+    xml_string = etree.tostring(tree, xml_declaration=True, encoding='UTF-8', pretty_print=module.params['pretty_print'])
+    changed = False
+
+    if module.params['file']:
+        xml_file = os.path.expanduser(module.params['file'])
+
+        xml_content = open(xml_file)
+        try:
+            if xml_string != xml_content.read():
+                changed = True
+                tree.write(xml_file, xml_declaration=True, encoding='UTF-8', pretty_print=module.params['pretty_print'])
+        finally:
+            xml_content.close()
+
+        module.exit_json(changed=changed)
+
+    if module.params['xmlstring']:
+        if xml_string != module.params['xmlstring']:
+            changed = True
+
+        module.exit_json(changed=changed, xmlstring=xml_string)
+
+def get_element_text(module, tree, xpath, namespaces):
+    if not is_node(tree, xpath, namespaces):
+        abort(module, "Xpath " + xpath + " does not reference a node!")
+
+    elements = []
+    for element in tree.xpath(xpath, namespaces=namespaces):
+        elements.append({element.tag: element.text})
+
+    finish(module, tree, xpath, namespaces, changed=False, msg=len(elements), hitcount=len(elements), matches=elements)
+
+def get_element_attr(module, tree, xpath, namespaces):
+    if not is_node(tree, xpath, namespaces):
+        abort(module, "Xpath " + xpath + " does not reference a node!")
+
+    elements = []
+    for element in tree.xpath(xpath, namespaces=namespaces):
+        child = {}
+        for key in element.keys():
+            value = element.get(key)
+            child.update({key: value})
+        elements.append({element.tag: child})
+
+    finish(module, tree, xpath, namespaces, changed=False, msg=len(elements), hitcount=len(elements), matches=elements)
+
+def child_to_element(module, child):
+    ch_type = type(child)
+    if ch_type == str or ch_type == unicode:
+        return etree.Element(child)
+    elif ch_type == dict:
+        if len(child) > 1:
+            abort(module, "Can only create children from hashes with one key")
+
+        (key, value) = child.items()[0]
+        if type(value) == dict:
+            children = value.pop('_', None)
+
+            node = etree.Element(key, value)
+
+            if children is not None:
+                if type(children) != list:
+                    abort(module, "Invalid children type: %s, must be list." % str(type(children)))
+
+                subnodes = children_to_nodes(module, children)
+                node.extend(subnodes)
+        else:
+            node = etree.Element(key)
+            node.text = value
+        return node
+    else:
+        abort(module, "Invalid child type: %s. Children must be either strings or hashes." % str(ch_type))
+
+def children_to_nodes(module=None, children=[]):
+    """turn a str/hash/list of str&hash into a list of elements"""
+    return [child_to_element(module, child) for child in children]
+
+def abort(m, msg):
+    m.fail_json(msg=msg)
```

Comment:
Most Ansible modules (all the ones I can think of) are written so that functions are defined before they are called.  So this helper function and finish should likely be near the top of the module file.